File: packages/client-swift/Sources/SpriteCoreClient/Ticker.swift
import Foundation

/// Timing abstraction for frame advancement. The default implementation uses
/// `Task.sleep`; tests inject a fake ticker that advances virtual time.
public protocol Ticker: Sendable {
    func delay(ms: Int) async throws
}

public struct SystemTicker: Ticker {
    public init() {}
    public func delay(ms: Int) async throws {
        try await Task.sleep(nanoseconds: UInt64(ms) * 1_000_000)
    }
}

File: packages/client-js/src/ticker.ts
/**
 * Timing abstraction for frame advancement. The default implementation uses
 * `setTimeout`; tests inject a fake ticker that advances virtual time. The
 * Kotlin + Swift ports expose the same seam.
 */
export interface Ticker {
  delay(ms: number): Promise<void>;
}

export class SystemTicker implements Ticker {
  delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
